MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c

```diff
@@ -839,103 +1203,225 @@ NvmeControllerInit (
+  // The number of queues allocated is returned and stored in the controller private data structure
+  // using the NumberOfDataQueuePairs field.
   //
-  Status = WriteNvmeAdminQueueAttributes (Private, &Aqa);
+  Status = NvmeSetFeaturesNumberOfQueues (Private, NVME_MAX_QUEUES - 1);
 
   if (EFI_ERROR (Status)) {
     return Status;
   }
 
+  // MU_CHANGE [END] - Request Number of Queues from Controller
+
+  // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
   //
-  // Program admin submission queue address.
+  // Allocate Data Queues - note we are assuming the queue entry sizes are the same as the admin queue entry sizes for the sake of memory allocation.
+  // The identify controller data tells us in SQES and CQES what the controller's minimum and maximum queue entry sizes are. We haven't used this before since we
+  // use the spec-defined minimum queue entry sizes.
+  // We are also allocating based on the admin defined queue sizes in number of entries.
+  // Some scenarios may use different queue sizes, currently we only see the case where the driver needs IO queue sizes <= admin queue sizes. So this allocation should be sufficient.
+  // We may want to explore a more dynamic allocation in the future.
   //
-  Status = WriteNvmeAdminSubmissionQueueBaseAddress (Private, &Asq);
+  for (Index = 1; Index <= Private->NumberOfDataQueuePairs; Index++) {
+    Private->SqData[Index].NumberOfEntries = Private->SqData[0].NumberOfEntries;
+    Private->CqData[Index].NumberOfEntries = Private->CqData[0].NumberOfEntries;
+    Private->SqData[Index].EntrySize       = Private->SqData[0].EntrySize;
+    Private->CqData[Index].EntrySize       = Private->CqData[0].EntrySize;
+  }
+
+  // Using the first data queue size for the number of pages required for the data queues
+  QueuePairPageCount = EFI_SIZE_TO_PAGES (Private->SqData[1].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[1].EntrySize))
+                       + EFI_SIZE_TO_PAGES (Private->CqData[1].NumberOfEntries * (UINTN)LShiftU64 (2, Private->CqData[1].EntrySize));
 
+  Status = PciIo->AllocateBuffer (
+                    PciIo,
+                    AllocateAnyPages,
+                    EfiBootServicesData,
+                    QueuePairPageCount * Private->NumberOfDataQueuePairs,
+                    (VOID **)&Private->DataQueueBuffer,
+                    0
+                    );
   if (EFI_ERROR (Status)) {
     return Status;
   }
 
+  Bytes  = EFI_PAGES_TO_SIZE (QueuePairPageCount * Private->NumberOfDataQueuePairs);
+  Status = PciIo->Map (
+                    PciIo,
+                    EfiPciIoOperationBusMasterCommonBuffer,
+                    Private->DataQueueBuffer,
+                    &Bytes,
+                    &MappedAddr,
+                    &Private->DataQueueMapping
+                    );
+
+  if (EFI_ERROR (Status) || (Bytes != EFI_PAGES_TO_SIZE (QueuePairPageCount * Private->NumberOfDataQueuePairs))) {
+    return Status;

```

Comment:
Same here, if the nvme controller isn't able to map the buffer that was allocated, will the caller code know about freeing the DataQueueBuffer?

Reply:
Same as other response ^

Reply:
same as above response.